script/vsts/platforms/templates/preparation.yml

```diff
@@ -20,9 +20,17 @@ steps:
     condition: eq(variables['Agent.OS'], 'Linux')
 
   # Common
+  - bash: |
+      if [ "${BUILD_ARCH}" = "x86" ]; then \
+        echo "##vso[task.setvariable variable=SHOULD_FORCE_32_BIT_NODE]true"
+      fi
+    displayName: Set SHOULD_FORCE_32_BIT_NODE based on BUILD_ARCH
+    # Templates evaluate too early for the matrix variable "BUILD_ARCH" to be available. Scripts evaluate at runtime, so "BUILD_ARCH" will be available.
```

Comment:
This sentence is not needed, and it a little inaccurate. Matrix variables (BUILD_ARCH) are calculated before running anything and so are available everywhere. It is not related to templates, scripts, etc.

https://github.com/atom/atom/blob/9e172191c08bd5b7e330b984abee774814a534df/script/vsts/platforms/windows.yml#L10

```suggestion
```

Reply:
You are incorrect about this. This is the documentation:

> To turn a pipeline into a run, Azure Pipelines goes through several steps in this order:
> 
> - 1. First, expand [templates](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops) and evaluate [template expressions](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops).
> [ ... ]
> - 4. For each job selected to run, expand [multi-configs](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops#parallelexec) (`strategy: matrix` or `strategy: parallel` in YAML) into multiple runtime jobs.
> [ ... ]
> 
> This ordering helps answer a common question: why can't I use certain variables in my template parameters? Step 1, template expansion, operates solely on the text of the YAML document. Runtime variables don't exist during that step. After step 1, template parameters have been completely resolved and no longer exist.

From here: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/runs?view=azure-devops#process-the-pipeline

---

I have tested many variations of this. The way I put in this pull request works, as do the "Alternate Designs" I mention in the body text of my pull request.

It is difficult to explain in a brief comment in the yml.

Reply:
I could change it to this, which is more precise:

```diff
-    # Templates evaluate too early for the matrix variable "BUILD_ARCH" to be available. Scripts evaluate at runtime, so "BUILD_ARCH" will be available.
+    # Templates evaluate too early for the matrix variable "BUILD_ARCH" to be available in ${{ template expressions }}.
+    # Scripts are run at runtime, so this script will make "SHOULD_FORCE_32_BIT_NODE" available to the following NodeTool task.
```

Reply:
The comment is correct now. 

 I still do not see why you are writing this so much verbose in the code. Better approach is to just put a link to the [Variable Definition section of Azure docs](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#what-syntax-should-i-use) on top of the YAML code, so everyone can read an learn about all the aspects of variable definition.

I have shortened the link for you: 
https://bit.ly/3dJk31m